### PR TITLE
New plugin: redfish

### PR DESCRIPTION
**Pull request by mmozejkx**
This is a C plugin that collects sensor data using Redfish protocol. 
It can get multiple resources from different hardware that support Redfish.

NOTE: Version of this code is an early drop.
Any feedback appreciated.

ChangeLog: Redfish plugin: Collect sensor data from Redfish endpoints.

[edit @kkepka:] added changelog

### CODE REVIEW

File: src/redfish.c [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+    if (ret != 0) {
+      ERROR(PLUGIN_NAME ": Something went wrong processing the service named \
+            \"%s\"",
+            service->name);
+      goto free_string_query;
+    }
+  }
+
+  llentry_t *entry = llentry_create(service->name, service);
+  if (entry != NULL)
```
Comment (octo) [inline src/redfish.c]: Handle error (and bail out).

File: src/redfish.c [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+
+  return 0;
+
+free_string_query:
+  sfree(service->name);
+  sfree(service->host);
+  sfree(service->user);
+  sfree(service->passwd);
+  sfree(service->token);
+  for (int j = 0; j < service->queries_num; j++)
```
Comment (octo) [inline src/redfish.c]: `strarray_free()` does this for you.

File: src/redfish.c [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+            service->name);
+      return -EINVAL;
+    }
+    if (service->queries_num == 0)
+      WARNING(PLUGIN_NAME ": Service \"%s\" does not have queries",
+              service->name);
+
+    for (int i = 0; i < service->queries_num; i++) {
+      redfish_query_t *query_query;
+      bool found = false;
```
Comment (octo) [inline src/redfish.c]: Please use standard C bools, i.e. `_Bool`. Also use `0` instead of `false`.

File: src/redfish.c [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+      char *key;
+      c_avl_iterator_t *query_iter = c_avl_get_iterator(ctx->queries);
+      while (c_avl_iterator_next(query_iter, (void **)&key,
```
Comment (octo) [inline src/redfish.c]: Cast to `(void *)` to improve portability.

File: src/redfish.c [hunk 5]
```diff
@@ -0,0 +1,891 @@
...
+                                 (void **)&query_query) == 0 &&
+             !found) {
+        if (query_query->name != NULL && service->queries[i] != NULL &&
+            strcmp(query_query->name, service->queries[i]) == 0) {
+          found = true;
```
Comment (octo) [inline src/redfish.c]: s/true/1/

File: src/redfish.c [hunk 6]
```diff
@@ -0,0 +1,891 @@
...
+      c_avl_iterator_destroy(query_iter);
+    }
+  }
+
+  c_avl_iterator_t *queries_iter = c_avl_get_iterator(ctx->queries);
+  char *key;
+  redfish_query_t *query;
+
+  /* Query validation */
+  while (c_avl_iterator_next(queries_iter, (void **)&key, (void **)&query) ==
```
Comment (octo) [inline src/redfish.c]: Cast to `(void *)`

File: src/redfish.c [hunk 7]
```diff
@@ -0,0 +1,891 @@
...
+    break;
+  default:
+    ERROR(PLUGIN_NAME ": Invalid data set type. Cannot convert value");
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+static void redfish_process_payload(bool success, unsigned short http_code,
```
Comment (octo) [inline src/redfish.c]: `bool` → `_Bool`

File: src/redfish.c [hunk 8]
```diff
@@ -0,0 +1,891 @@
...
+                                    redfishPayload *payload, void *context) {
+  if (success == false) {
```
Comment (octo) [inline src/redfish.c]: `if (!success)`

File: src/redfish.c [hunk 9]
```diff
@@ -0,0 +1,891 @@
...
+    WARNING(PLUGIN_NAME ": Query has failed, HTTP code = %u\n", http_code);
+    return;
+  }
+  redfish_payload_ctx_t *res_serv = (redfish_payload_ctx_t *)context;
+  redfish_service_t *serv = res_serv->service;
+  if (payload) {
```
Comment (octo) [inline src/redfish.c]: Please handle error and out-dent the normal flow. I.e.:

```c
if (!payload) {
  WARNING();
  return;
}
/* normal case goes here */
```

File: src/redfish.c [hunk 10]
```diff
@@ -0,0 +1,891 @@
...
+      for (llentry_t *llprop = llist_head(res->properties); llprop != NULL;
+           llprop = llprop->next) {
+        redfish_property_t *prop = (redfish_property_t *)llprop->value;
+        if (json_array == NULL) {
+          ERROR("Could not find resource \"%s\"", res->name);
+          continue;
+        }
+
+        /* Iterating through array of sensor(s) */
+        for (int i = 0; i < json_array_size(json_array); i++) {
```
Comment (octo) [inline src/redfish.c]: This is the third nested loop. Please break this function up into smaller functiosn that are easier to comprehend.

File: src/redfish.c [hunk 11]
```diff
@@ -0,0 +1,891 @@
...
+    json_t *json_array;
+    for (llentry_t *llres = llist_head(res_serv->resources); llres != NULL;
+         llres = llres->next) {
+      redfish_resource_t *res = (redfish_resource_t *)llres->value;
+      json_array = json_object_get(payload->json, res->name);
+      for (llentry_t *llprop = llist_head(res->properties); llprop != NULL;
+           llprop = llprop->next) {
+        redfish_property_t *prop = (redfish_property_t *)llprop->value;
+        if (json_array == NULL) {
+          ERROR("Could not find resource \"%s\"", res->name);
```
Comment (octo) [inline src/redfish.c]: FYI: ERRORS in this function are missing the `PLUGIN_NAME` prefix.

File: src/redfish.c [hunk 12]
```diff
@@ -0,0 +1,891 @@
...
+        for (int i = 0; i < json_array_size(json_array); i++) {
+          json_t *member_id;
+          json_t *object;
+          json_t *item = json_array_get(json_array, i);
+          if (item == NULL) {
+            ERROR("Failure retrieving array member for resource \"%s\"",
+                  res->name);
+            continue;
+          }
+          object = json_object_get(item, prop->name);
```
Comment (octo) [inline src/redfish.c]: Please declare `object` here.

File: src/redfish.c [hunk 13]
```diff
@@ -0,0 +1,891 @@
...
+static void redfish_process_payload(bool success, unsigned short http_code,
+                                    redfishPayload *payload, void *context) {
+  if (success == false) {
+    WARNING(PLUGIN_NAME ": Query has failed, HTTP code = %u\n", http_code);
+    return;
+  }
+  redfish_payload_ctx_t *res_serv = (redfish_payload_ctx_t *)context;
+  redfish_service_t *serv = res_serv->service;
+  if (payload) {
+    json_t *json_array;
```
Comment (octo) [inline src/redfish.c]: Does this really need to live longer than the inner loop? It's assigned unconditionally inside the loop, which feels leaky to me.

File: src/redfish.c [hunk 14]
```diff
@@ -0,0 +1,891 @@
...
+          else {
+            /* Retrieving MemberId of sensor */
+            member_id = json_object_get(item, "MemberId");
+            if (member_id == NULL) {
+              ERROR("Failed to get MemberId for property \"%s\" in resource "
+                    "\"%s\"",
+                    prop->name, res->name);
+              continue;
+            }
+            char type_instance[sizeof(v1.type_instance)];
```
Comment (octo) [inline src/redfish.c]: Why don't you let `snprintf()` write into `v1.type_instance` directly?

File: src/redfish.c [hunk 15]
```diff
@@ -0,0 +1,891 @@
...
+    WARNING("Failed to get payload for service name \"%s\"", serv->name);
+  }
+}
+
+static int redfish_read(__attribute__((unused)) user_data_t *ud) {
+  for (llentry_t *le = llist_head(ctx->services); le != NULL; le = le->next) {
+    redfish_service_t *service = (redfish_service_t *)le->value;
+    for (llentry_t *le = llist_head(service->query_ptrs); le != NULL;
+         le = le->next) {
+      redfish_query_t *query = (redfish_query_t *)le->value;
```
Comment (octo) [inline src/redfish.c]: Optional: unnecessary cast.

File: src/redfish.c [hunk 16]
```diff
@@ -0,0 +1,891 @@
...
+      json_decref(json_array);
+    }
+  } else {
+    WARNING("Failed to get payload for service name \"%s\"", serv->name);
+  }
+}
+
+static int redfish_read(__attribute__((unused)) user_data_t *ud) {
+  for (llentry_t *le = llist_head(ctx->services); le != NULL; le = le->next) {
+    redfish_service_t *service = (redfish_service_t *)le->value;
```
Comment (octo) [inline src/redfish.c]: Optional: unnecessary cast.

File: src/redfish.c [hunk 17]
```diff
@@ -0,0 +1,891 @@
...
+  for (llentry_t *le = llist_head(ctx->services); le; le = le->next) {
+    redfish_service_t *service = (redfish_service_t *)le->value;
+
+    cleanupServiceEnumerator(service->redfish);
+    for (size_t i = 0; i < service->queries_num; i++)
+      sfree(service->queries[i]);
+
+    llist_destroy(service->query_ptrs);
+
+    sfree(service->name);
```
Comment (octo) [inline src/redfish.c]: I've seen this list of sfree statements for various service members before. Maybe put that into a `service_destroy()` function?

File: src/redfish.c [hunk 18]
```diff
@@ -0,0 +1,891 @@
...
+  return 0;
+}
+
+static int redfish_cleanup(void) {
+  for (llentry_t *le = llist_head(ctx->services); le; le = le->next) {
+    redfish_service_t *service = (redfish_service_t *)le->value;
+
+    cleanupServiceEnumerator(service->redfish);
+    for (size_t i = 0; i < service->queries_num; i++)
+      sfree(service->queries[i]);
```
Comment (octo) [inline src/redfish.c]: Optional: Consider using `strarray_free()`.

File: src/redfish.c [hunk 19]
```diff
@@ -0,0 +1,891 @@
...
+    sfree(service);
+  }
+  llist_destroy(ctx->services);
+
+  c_avl_iterator_t *i = c_avl_get_iterator(ctx->queries);
+
+  char *key;
+  redfish_query_t *query;
+
+  while (c_avl_iterator_next(i, (void **)&key, (void **)&query) == 0) {
```
Comment (octo) [inline src/redfish.c]: Cast to `(void *)`

File: src/redfish.c (hunk 3 above)
Comment (mmozejkx) [reply src/redfish.c]: I'm surprised you're asking to change it because previously we were switching from `_Bool` to `bool` and changing `0` to `false` :)

**Comment by mmozejkx** [comment]
@octo Thanks for your feedback. We're working on creating a patch and it should be upstreamed soon.

Regards,
Marcin

**Comment by mkobyli** [comment]
Hi @octo,
We updated our PR.
I'll be happy for your review. Please let me know, if PR need any fix.
Regards,
Michal

File: src/redfish.c (hunk 3 above)
Comment (mkobyli) [reply src/redfish.c]: @octo Is this a valid or it need a fix?

**Comment by sunkuranganath** [comment]
@mkobyli Could you please resolve the conflicts?

File: Makefile.am [hunk 20]
```diff
@@ -1617,6 +1617,26 @@ protocols_la_LDFLAGS = $(PLUGIN_LDFLAGS)
...
 
+if BUILD_PLUGIN_REDFISH
+pkglib_LTLIBRARIES += redfish.la
+redfish_la_SOURCES = src/redfish.c
+redfish_la_CPPFLAGS = $(AM_CPPFLAGS) $(BUILD_WITH_LIBREDFISH_CPPFLAGS)
+redfish_la_LDFLAGS = $(PLUGIN_LDFLAGS) $(BUILD_WITH_LIBREDFISH_LDFLAGS)
+redfish_la_LIBADD = $(BUILD_WITH_LIBREDFISH_LIBS) -lredfish
+
+test_plugin_redfish_SOURCES = src/redfish_test.c \
+                              src/daemon/utils_avltree.c \
```
Comment (kkepka) [inline Makefile.am]: Please followup on [changed location of utils_avltree](https://github.com/collectd/collectd/commit/6378ec288f34ff250b2971a1452338a2b34c240a#diff-bd2cd6529005fa0569e8cfa99f0534dc).

**Comment by kkepka** [comment]
my apologies for "merge ladder", should squash it..

**Pull request merged**